### PR TITLE
adding support for kwargs in `load_state`

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Fixes: https://huggingface.slack.com/archives/C01Q6JPP6NA/p1672362244049489. Adds `kwargs` keyword arguments to the `load_state` method of accelerate.

### CODE REVIEW

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -1717,6 +1717,8 @@ def load_state(self, input_dir: str):
         Args:
             input_dir (`str` or `os.PathLike`):
                 The name of the folder all relevant weights and states were saved in.
+            model_load_kwargs (`dict`): Additional keyword arguments required for loading model, e.g.,
+            deepspeed's optional args for `load_checkpoint`.
```
Comment (sgugger) [inline src/accelerate/accelerator.py:1721]:
```suggestion
            model_load_kwargs (`dict`):
                Additional keyword arguments required for loading model, e.g., deepspeed's optional args for `load_checkpoint`.
```

**Review by sgugger — APPROVED** [review]
Thanks for working on this!

**Comment by pacman100** [comment]
cc @stas00

File: src/accelerate/accelerator.py [hunk 2]
```diff
@@ -1704,7 +1704,7 @@ def save_state(self, output_dir: str = None):
         self.project_configuration.iteration += 1
         return save_location
 
-    def load_state(self, input_dir: str):
+    def load_state(self, input_dir: str, **model_load_kwargs):
```
Comment (muellerzr) [inline src/accelerate/accelerator.py:1707]:
```suggestion
    def load_state(self, input_dir: str, **load_func_kwargs):
```

I think we should rename this to `load_func_kwargs` which is a bit more explicit in what these are actually going towards.

File: src/accelerate/accelerator.py [hunk 3]
```diff
@@ -1717,6 +1717,9 @@ def load_state(self, input_dir: str):
         Args:
             input_dir (`str` or `os.PathLike`):
                 The name of the folder all relevant weights and states were saved in.
+            model_load_kwargs (`dict`):
+                Additional keyword arguments required for loading model, e.g., deepspeed's optional args for
+                `load_checkpoint`.
```
Comment (muellerzr) [inline src/accelerate/accelerator.py:1722]:
```suggestion
                Additional keyword arguments required for loading model which are passed to the underlying load function, such as optional arguments for `DeepSpeed`'s `load_checkpoint` function.
```

Let's be a bit more explicit here

File: src/accelerate/checkpointing.py [hunk 4]
```diff
@@ -126,12 +126,13 @@ def load_accelerator_state(input_dir, models, optimizers, schedulers, process_in
             The current process index in the Accelerator state
         scaler (`torch.cuda.amp.GradScaler`, *optional*):
             An optional *GradScaler* instance to load
+        model_load_kwargs (`dict`): Additional arguments to pass to the model's `load_state_dict` method.
```
Comment (muellerzr) [inline src/accelerate/checkpointing.py:129]: Also bring the earlier suggestions into this section (including the spacing Sylvain suggested earlier)

**Review by muellerzr — APPROVED** [review]
Nice addition! Just a few requests for naming and documentation clarity

**Review by muellerzr — APPROVED** [review]
Let's also try to add a test or three (since we have FSDP, DeepSpeed, and Megatron) to ensure that everything is working properly if possible :)

File: src/accelerate/accelerator.py (hunk 2 above)
Comment (pacman100) [reply src/accelerate/accelerator.py:1707]: hello, but `load_func` can mean optimizer/scheduler loading too, and hence chose this name.

Comment (muellerzr) [reply src/accelerate/accelerator.py:1707]: True, `load_model_func_kwargs` perhaps then as a medium

**Comment by stas00** [comment]
Thank you very much for adding this functionality, Sourab et al.

I don't really need it right now, but do you want to add the `save_checkpoint` kwargs for symmetry? e.g. deepspeed's is:

```
    def save_checkpoint(self, save_dir, tag=None, client_state={}, save_latest=True):
        r"""Save training checkpoint
        Arguments:
            save_dir: Required. Directory for saving the checkpoint
            tag: Optional. Checkpoint tag used as a unique identifier for the checkpoint, global step is
                used if not provided. Tag name must be the same across all ranks.
            client_state: Optional. State dictionary used for saving required training states in the client code.
            save_latest: Optional. Save a file 'latest' pointing to the latest saved checkpoint.
```

but as I said this is totally a suggestion and not a need.

Thank you!

File: src/accelerate/accelerator.py (hunk 2 above)
Comment (pacman100) [reply src/accelerate/accelerator.py]: Done

File: src/accelerate/accelerator.py [hunk 5]
```diff
@@ -1634,6 +1634,9 @@ def save_state(self, output_dir: str = None):
         Args:
             output_dir (`str` or `os.PathLike`):
                 The name of the folder to save all relevant weights and states.
+            save_model_func_kwargs (`dict`, *optional*):
+                Additional keyword arguments for saving model which can be passed to the underlying save function, such
+                as optional arguments for `DeepSpeed`'s `save_checkpoint` function.
```
Comment (stas00) [inline src/accelerate/accelerator.py:1639]:
```suggestion
                as optional arguments for DeepSpeed's `save_checkpoint` function.
```

either  `deepspeed` (package) or  DeepSpeed (name / no backquotes)

File: src/accelerate/accelerator.py [hunk 6]
```diff
@@ -1717,6 +1720,9 @@ def load_state(self, input_dir: str):
         Args:
             input_dir (`str` or `os.PathLike`):
                 The name of the folder all relevant weights and states were saved in.
+            load_model_func_kwargs (`dict`, *optional*):
+                Additional keyword arguments for loading model which can be passed to the underlying load function,
+                such as optional arguments for `DeepSpeed`'s `load_checkpoint` function.
```
Comment (stas00) [inline src/accelerate/accelerator.py:1725]:
```suggestion
                such as optional arguments for DeepSpeed's `load_checkpoint` function.
```

**Review by stas00 — APPROVED** [review]
Thank you, Sourab - looking great.

I left one small suggestion (2x)

**Review by muellerzr — APPROVED** [review]
Great!
